integration/v3_grpc_test.go

```diff
@@ -491,3 +491,86 @@ func TestV3WatchCancel(t *testing.T) {
 
 	clus.Terminate(t)
 }
+
+// TestV3WatchMultiple tests multiple watchers on the same key
+// and one watcher with matching prefix. It first puts the key
+// that matches all watchers, and another key that matches only
+// one watcher to test if it receives expected events.
+func TestV3WatchMultiple(t *testing.T) {
+	clus := newClusterGRPC(t, &clusterConfig{size: 3})
+	wAPI := pb.NewWatchClient(clus.RandConn())
+	kvc := pb.NewKVClient(clus.RandConn())
+
+	wStream, errW := wAPI.Watch(context.TODO())
+	if errW != nil {
+		t.Fatalf("wAPI.Watch error: %v", errW)
+	}
+
+	go func() {
+		if _, err := kvc.Put(context.TODO(), &pb.PutRequest{Key: []byte("foo"), Value: []byte("bar")}); err != nil {
+			t.Fatalf("couldn't put key (%v)", err)
+		}
+	}()
+
+	keysToWatch := []string{"fo", "foo", "foo", "foo", "foo"}
+	keysToReceive := []string{"foo", "foo", "foo", "foo", "foo"}
+
+	for i := range keysToWatch {
+		go func(key string) {
+			if err := wStream.Send(&pb.WatchRequest{CreateRequest: &pb.WatchCreateRequest{Prefix: []byte(key)}}); err != nil {
+				t.Errorf("wStream.Send error: %v", err)
+			}
+		}(keysToWatch[i])
+	}
+
+	keysReceived := []string{}
+	for len(keysToWatch) != len(keysReceived) {
+		wresp, err := wStream.Recv()
+		if err != nil {
+			t.Errorf("wStream.Recv error: %v", err)
+		}
+		if wresp.Created {
+			continue
+		}
+		for _, ev := range wresp.Events {
+			keysReceived = append(keysReceived, string(ev.Kv.Key))
+		}
+	}
+
+	if !reflect.DeepEqual(keysToReceive, keysReceived) {
+		t.Errorf("unexpected keys received %#q, want %#q", keysReceived, keysToWatch)
+	}
+
+	if _, err := kvc.Put(context.TODO(), &pb.PutRequest{Key: []byte("fo"), Value: []byte("bar")}); err != nil {
+		t.Fatalf("couldn't put key (%v)", err)
+	}
+
+	wresp, err := wStream.Recv()
+	if err != nil {
+		t.Errorf("wStream.Recv error: %v", err)
+	}
+	if len(wresp.Events) != 1 {
+		t.Errorf("len(wresp.Events) got = %d, want = 1", len(wresp.Events))
```

Comment:
Please be careful on this one. if len(wresp.Event) = 0, we might panic at line 555. We have to fatal here.